src/shaders/_prelude_fog.fragment.glsl

```diff
@@ -15,45 +25,52 @@ float fog_sky_blending(vec3 camera_dir) {
     return u_fog_opacity * exp(-3.0 * t * t);
 }
 
-float fog_opacity(vec3 pos) {
-    float depth = length(pos);
-    float start = u_fog_range.x;
-    float end = u_fog_range.y;
-
-    // The fog is not physically accurate, so we seek an expression which satisfies a
-    // couple basic constraints:
-    //   - opacity should be 0 at the near limit
-    //   - opacity should be 1 at the far limit
-    //   - the onset should have smooth derivatives to avoid a sharp band
-    // To this end, we use an (1 - e^x)^n, where n is set to 3 to ensure the
-    // function is C2 continuous at the onset. The fog is about 99% opaque at
-    // the far limit, so we simply scale it and clip to achieve 100% opacity.
-    // https://www.desmos.com/calculator/3taufutxid
-    const float decay = 5.5;
-    float falloff = max(0.0, 1.0 - exp(-decay * (depth - start) / (end - start)));
+// This function gives the fog opacity when strength is 1. Otherwise it's multiplied
+// by a smoothstep to a power to decrease the amount of fog relative to haze.
+// This function much match src/style/fog.js
+// See: https://www.desmos.com/calculator/3taufutxid
+float fog_opacity(float t) {
+    const float decay = 6.0;
+    float falloff = 1.0 - min(1.0, exp(-decay * t));
 
     // Cube without pow()
     falloff *= falloff * falloff;
 
     // Scale and clip to 1 at the far limit
-    falloff = min(1.0, 1.00747 * falloff);
+    return u_fog_opacity * min(1.0, 1.00747 * falloff);
+}
+
+float fog_opacity (vec3 pos) {
+    return fog_opacity((length(pos) - u_fog_range.x) / (u_fog_range.y - u_fog_range.x));
+}
+
+vec3 fog_apply(vec3 color, vec3 pos) {
+    // Map [near, far] to [0, 1]
+    float t = (length(pos) - u_fog_range.x) / (u_fog_range.y - u_fog_range.x);
 
-    return falloff * u_fog_opacity;
+    float haze_opac = fog_opacity(pos);
```

Comment:
Could you rearrange this code to be wrapping the shader expresion `(length(pos) - u_fog_range.x) / (u_fog_range.y - u_fog_range.x)`, otherwise it's evaluated twice: for `t` and within the call to `fog_opacity` (maybe in a function or by passing `t` to `fog_opacity` directly?).

Reply:
The short answer is not cleanly, no, I don't think so. The statement is duplicated specifically so that the operations are not repeated. I instead added a note pointing out that fog_opacity(vec3) is only rarely called (e.g. heatmap) while all other fog opacity computations go through `fog_opacity(float)`.